File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
class Solution {
public:
   int minEatingSpeed(vector<int>& piles, int h) {
        int sz = piles.size(); 
        int j = *max_element(piles.begin(), piles.end());
        // small optimization : just take max from vector if vector.size() == h 
        if(sz == h) return j;
        int i = 1;
        while(i < j){
            int temp_h = 0;
            int mid = i + ((j - i) / 2);
            for(int k = 0; k < sz; ++k)
                temp_h += ceil((double)piles[k] / mid);
            
            if (temp_h > h) i = mid + 1;
            else j = mid;
        }
        return i;
    }
};

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.rs
impl Solution {
    pub fn min_eating_speed(piles: Vec<i32>, h: i32) -> i32 {
        let mut left = 1;
        let mut right = *piles.iter().max().unwrap(); // maximum pile size

        while left < right {
            let mid = left + (right - left) / 2;
            if Solution::calculate_hours(&piles, mid) <= h {
                right = mid; // try smaller speeds
            } else {
                left = mid + 1; // try larger speeds
            }
        }
        left
    }

    // Helper function to calculate the total number of hours Koko needs at speed `k`
    fn calculate_hours(piles: &Vec<i32>, k: i32) -> i32 {
        let mut hours = 0;
        for &pile in piles {
            hours += (pile + k - 1) / k; // this is equivalent to ceil(pile / k)
        }
        hours
    }
}